{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import malariagen_data\n",
    "import allel\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import numpy as np\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    function drop(id) {\n      const view = Bokeh.index.get_by_id(id)\n      if (view != null) {\n        view.model.document.clear()\n        Bokeh.index.delete(view)\n      }\n    }\n\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n\n    // Clean up Bokeh references\n    if (id != null) {\n      drop(id)\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim()\n            drop(id)\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.3.4.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "            <table class=\"malariagen-ag3\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\" colspan=\"2\">MalariaGEN Ag3 API client</th>\n",
       "                    </tr>\n",
       "                    <tr><td colspan=\"2\" style=\"text-align: left\">\n",
       "                        Please note that data are subject to terms of use,\n",
       "                        for more information see <a href=\"https://www.malariagen.net/data\">\n",
       "                        the MalariaGEN website</a> or contact support@malariagen.net.\n",
       "                        See also the <a href=\"https://malariagen.github.io/malariagen-data-python/v0.0.0.post782+5e502bf/Ag3.html\">Ag3 API docs</a>.\n",
       "                    </td></tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Storage URL\n",
       "                        </th>\n",
       "                        <td>gs://vo_agam_release/</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Data releases available\n",
       "                        </th>\n",
       "                        <td>3.0, 3.1, 3.10, 3.11, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Results cache\n",
       "                        </th>\n",
       "                        <td>/Users/dennistpw/Projects/drc_genomic_surveillance/data</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Cohorts analysis\n",
       "                        </th>\n",
       "                        <td>20240717</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            AIM analysis\n",
       "                        </th>\n",
       "                        <td>20220528</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Site filters analysis\n",
       "                        </th>\n",
       "                        <td>dt_20200416</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Software version\n",
       "                        </th>\n",
       "                        <td>malariagen_data 0.0.0.post782+5e502bf</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Client location\n",
       "                        </th>\n",
       "                        <td>Queensland, Australia</td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        "
      ],
      "text/plain": [
       "<MalariaGEN Ag3 API client>\n",
       "Storage URL             : gs://vo_agam_release/\n",
       "Data releases available : 3.0, 3.1, 3.10, 3.11, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9\n",
       "Results cache           : /Users/dennistpw/Projects/drc_genomic_surveillance/data\n",
       "Cohorts analysis        : 20240717\n",
       "AIM analysis            : 20220528\n",
       "Site filters analysis   : dt_20200416\n",
       "Software version        : malariagen_data 0.0.0.post782+5e502bf\n",
       "Client location         : Queensland, Australia\n",
       "---\n",
       "Please note that data are subject to terms of use,\n",
       "for more information see https://www.malariagen.net/data\n",
       "or contact support@malariagen.net. For API documentation see \n",
       "https://malariagen.github.io/malariagen-data-python/v0.0.0.post782+5e502bf/Ag3.html"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ag3 = malariagen_data.Ag3(\n",
    "    results_cache=\"/Users/dennistpw/Projects/drc_genomic_surveillance/data/\",\n",
    "    pre=True\n",
    ")\n",
    "ag3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define all the sample sets we want to include and exclude based on Poppy's analyses\n",
    "drc_sample_sets = [\n",
    "    '1264-VO-CD-WATSENGA-VMF00161',\n",
    "    '1264-VO-CD-WATSENGA-VMF00164',\n",
    "]\n",
    "ref_sample_sets = [\n",
    "    # 'AG1000G-CD',\n",
    "    'AG1000G-CF',\n",
    "    'AG1000G-TZ',\n",
    "    'AG1000G-UG',\n",
    "#    'AG1000G-GA-A',\n",
    "#    'AG1000G-MZ',\n",
    "    'AG1000G-CM-A',\n",
    "    'AG1000G-BF-A',\n",
    "    '1273-VO-ZM-MULEBA-VMF00176',\n",
    "]\n",
    "sample_sets = drc_sample_sets + ref_sample_sets\n",
    "\n",
    "locations_blacklist = [\n",
    "    'Mayos',\n",
    "    'Daiguene',\n",
    "    'Gbadolite',\n",
    "    'Muheza',\n",
    "]\n",
    "\n",
    "samples_blacklist = [\n",
    "    \"AB0203-C\",\n",
    "    \"AB0177-C\",\n",
    "    \"AN0085-C\",\n",
    "    \"AN0107-C\",\n",
    "    \"AC0223-C\",\n",
    "    \"AC0240-C\",\n",
    "    \"AB0207-C\",\n",
    "    \"AB0158-Cx\",\n",
    "]\n",
    "\n",
    "drc_samples_blacklist = ['VBS48708-6367STDY9888400', 'VBS48725-6367STDY9888417']\n",
    "\n",
    "cat_ordered = [\n",
    "    'Burkina Faso',\n",
    "    'Cameroon',\n",
    "    'Central African Republic',\n",
    "    'Uganda',\n",
    "    'Tanzania',\n",
    "    'Zambia',\n",
    "    'DRC Far North',\n",
    "    'DRC South East',\n",
    "    'DRC South West',\n",
    "    'DRC Forest'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "group\n",
       "Burkina Faso                 95\n",
       "Cameroon                     95\n",
       "Central African Republic     55\n",
       "DRC Far North                46\n",
       "DRC Forest (Tshopo)         118\n",
       "DRC South East              320\n",
       "DRC South West              187\n",
       "Tanzania                     32\n",
       "Uganda                      205\n",
       "Zambia                      201\n",
       "dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now define sample analysis pop from metadata\n",
    "\n",
    "drc_metadata = ag3.sample_metadata(sample_query=(\n",
    "        \"taxon == 'gambiae' and \"\n",
    "        f\"sample_set in {sample_sets} and \"\n",
    "        f\"location not in {locations_blacklist} and \"\n",
    "        f\"sample_id not in {samples_blacklist + drc_samples_blacklist}\"\n",
    "    ))\n",
    "\n",
    "# add coloring category\n",
    "drc_metadata.loc[:, \"group\"] = drc_metadata[\"country\"].copy()\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"Nord-Ubangi\"), \"group\"] = \"DRC Far North\"\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"Upper Uele\"), \"group\"] = \"DRC Far North\"\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"Kongo-Central\"), \"group\"] = \"DRC South West\"\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"Kinshasa\"), \"group\"] = \"DRC South West\"\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"Mai-Ndombe\"), \"group\"] = \"DRC South West\"\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"Sankuru\"), \"group\"] = \"DRC South West\"\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"South Kivu\"), \"group\"] = \"DRC South East\"\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"Central Kasai\"), \"group\"] = \"DRC South East\"\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"Tanganyika\"), \"group\"] = \"DRC South East\"\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"Haut-Katanga\"), \"group\"] = \"DRC South East\"\n",
    "drc_metadata.loc[(drc_metadata[\"admin1_name\"] == \"Tshopo\"), \"group\"] = \"DRC Forest (Tshopo)\"\n",
    "drc_metadata.groupby(\"group\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "selscan_list = []\n",
    "for pop in drc_metadata.group.unique():\n",
    "    sample_list = drc_metadata[drc_metadata.group == pop]['sample_id'].tolist()\n",
    "    for chrom in ag3.contigs:\n",
    "        h = ag3.h12_gwss(\n",
    "               contig=chrom,\n",
    "               sample_query=f'sample_id=={sample_list}',\n",
    "               cohort_size=20,\n",
    "               window_size=2000,\n",
    "           )\n",
    "        h = (*h, chrom)\n",
    "        h = (*h, pop)\n",
    "        selscan_list.append(h)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "#set up some plotting variables\n",
    "maxpos = []\n",
    "for i in [0,1,2,3,4]:\n",
    "    maxpos.append(selscan_list[i][0].max())\n",
    "total_max = sum(maxpos)\n",
    "column_widths = (maxpos / total_max)\n",
    "col_var_levels = ag3.contigs\n",
    "\n",
    "#import useful libs\n",
    "from matplotlib.ticker import FuncFormatter\n",
    "from matplotlib.gridspec import GridSpec\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "row_var_levels = drc_metadata.group.unique()\n",
    "rowvars_reordered = ['Burkina Faso','Cameroon', 'Central African Republic','DRC Far North','DRC Forest (Tshopo)', 'DRC South East','Zambia','Tanzania','Uganda']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert each tuple into a DataFrame, expanding shorter lists\n",
    "dfs = []\n",
    "for i, (pos, h12, chrom, pop) in enumerate(selscan_list):\n",
    "    max_len = len(pos)  # Determine the length to expand to\n",
    "    df = pd.DataFrame({\n",
    "        'pos': pos,\n",
    "        'h12': h12,\n",
    "        'chrom': chrom,  # Repeat chromosome to match length\n",
    "        'pop': pop       # Repeat population to match length\n",
    "    })\n",
    "    dfs.append(df)\n",
    "\n",
    "# Combine all DataFrames into one\n",
    "selscan_df = pd.concat(dfs, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('2R', '2L', '3R', '3L', 'X')"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_var_levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, (array([6.92930200e+03, 1.37947570e+04, 2.05072190e+04, ...,\n",
      "       6.13748295e+07, 6.14054660e+07, 6.14376831e+07]), array([0.08   , 0.05875, 0.07625, ..., 0.28   , 0.12875, 0.055  ]), '2R', 'DRC Forest (Tshopo)'))\n",
      "(1, (array([  106585.182 ,   184446.54  ,   221036.9535, ..., 49335563.836 ,\n",
      "       49342660.191 , 49349478.1905]), array([0.0375 , 0.39875, 0.0975 , ..., 0.1025 , 0.0975 , 0.075  ]), '2L', 'DRC Forest (Tshopo)'))\n",
      "(2, (array([9.59644600e+03, 1.86722940e+04, 2.45557020e+04, ...,\n",
      "       5.29666158e+07, 5.30451037e+07, 5.30983105e+07]), array([0.05125, 0.34875, 0.08875, ..., 0.1675 , 0.055  , 0.0725 ]), '3R', 'DRC Forest (Tshopo)'))\n",
      "(3, (array([   68036.8035,    82886.502 ,   111535.469 , ..., 41931997.7705,\n",
      "       41943264.259 , 41951070.9215]), array([0.055  , 0.07125, 0.0625 , ..., 0.57   , 0.57   , 0.61625]), '3L', 'DRC Forest (Tshopo)'))\n",
      "(4, (array([4.78364350e+03, 2.33791065e+04, 3.78083210e+04, ...,\n",
      "       2.42617551e+07, 2.42873805e+07, 2.43315226e+07]), array([0.54125, 0.2575 , 0.42625, ..., 0.73   , 0.15375, 0.10375]), 'X', 'DRC Forest (Tshopo)'))\n",
      "(5, (array([6.92930200e+03, 1.37947570e+04, 2.05072190e+04, ...,\n",
      "       6.13748295e+07, 6.14054660e+07, 6.14376831e+07]), array([0.11625, 0.065  , 0.135  , ..., 0.34625, 0.135  , 0.0725 ]), '2R', 'DRC South East'))\n",
      "(6, (array([  106585.182 ,   184446.54  ,   221036.9535, ..., 49335563.836 ,\n",
      "       49342660.191 , 49349478.1905]), array([0.04625, 0.43625, 0.17875, ..., 0.1225 , 0.11875, 0.09875]), '2L', 'DRC South East'))\n",
      "(7, (array([9.59644600e+03, 1.86722940e+04, 2.45557020e+04, ...,\n",
      "       5.29666158e+07, 5.30451037e+07, 5.30983105e+07]), array([0.04875, 0.19875, 0.06875, ..., 0.13375, 0.02875, 0.075  ]), '3R', 'DRC South East'))\n",
      "(8, (array([   68036.8035,    82886.502 ,   111535.469 , ..., 41931997.7705,\n",
      "       41943264.259 , 41951070.9215]), array([0.02875, 0.035  , 0.03375, ..., 0.34   , 0.2525 , 0.27   ]), '3L', 'DRC South East'))\n",
      "(9, (array([4.78364350e+03, 2.33791065e+04, 3.78083210e+04, ...,\n",
      "       2.42617551e+07, 2.42873805e+07, 2.43315226e+07]), array([0.8125 , 0.60875, 0.4425 , ..., 0.46875, 0.02625, 0.0525 ]), 'X', 'DRC South East'))\n",
      "(10, (array([6.92930200e+03, 1.37947570e+04, 2.05072190e+04, ...,\n",
      "       6.13748295e+07, 6.14054660e+07, 6.14376831e+07]), array([0.07   , 0.0475 , 0.1025 , ..., 0.35375, 0.0825 , 0.0775 ]), '2R', 'DRC South West'))\n",
      "(11, (array([  106585.182 ,   184446.54  ,   221036.9535, ..., 49335563.836 ,\n",
      "       49342660.191 , 49349478.1905]), array([0.02875, 0.3375 , 0.08375, ..., 0.08625, 0.20125, 0.12   ]), '2L', 'DRC South West'))\n",
      "(12, (array([9.59644600e+03, 1.86722940e+04, 2.45557020e+04, ...,\n",
      "       5.29666158e+07, 5.30451037e+07, 5.30983105e+07]), array([0.0375 , 0.22875, 0.0675 , ..., 0.115  , 0.035  , 0.07   ]), '3R', 'DRC South West'))\n",
      "(13, (array([   68036.8035,    82886.502 ,   111535.469 , ..., 41931997.7705,\n",
      "       41943264.259 , 41951070.9215]), array([0.03375, 0.07125, 0.07125, ..., 0.20375, 0.215  , 0.255  ]), '3L', 'DRC South West'))\n",
      "(14, (array([4.78364350e+03, 2.33791065e+04, 3.78083210e+04, ...,\n",
      "       2.42617551e+07, 2.42873805e+07, 2.43315226e+07]), array([0.47375, 0.4375 , 0.57   , ..., 0.55125, 0.03625, 0.03625]), 'X', 'DRC South West'))\n",
      "(15, (array([6.92930200e+03, 1.37947570e+04, 2.05072190e+04, ...,\n",
      "       6.13748295e+07, 6.14054660e+07, 6.14376831e+07]), array([0.06   , 0.03375, 0.0675 , ..., 0.2575 , 0.1425 , 0.055  ]), '2R', 'DRC Far North'))\n",
      "(16, (array([  106585.182 ,   184446.54  ,   221036.9535, ..., 49335563.836 ,\n",
      "       49342660.191 , 49349478.1905]), array([0.02875, 0.3175 , 0.04875, ..., 0.03875, 0.05125, 0.03375]), '2L', 'DRC Far North'))\n",
      "(17, (array([9.59644600e+03, 1.86722940e+04, 2.45557020e+04, ...,\n",
      "       5.29666158e+07, 5.30451037e+07, 5.30983105e+07]), array([0.03875, 0.1825 , 0.07875, ..., 0.12375, 0.0625 , 0.05625]), '3R', 'DRC Far North'))\n",
      "(18, (array([   68036.8035,    82886.502 ,   111535.469 , ..., 41931997.7705,\n",
      "       41943264.259 , 41951070.9215]), array([0.0375 , 0.0575 , 0.04875, ..., 0.5325 , 0.43125, 0.4975 ]), '3L', 'DRC Far North'))\n",
      "(19, (array([4.78364350e+03, 2.33791065e+04, 3.78083210e+04, ...,\n",
      "       2.42617551e+07, 2.42873805e+07, 2.43315226e+07]), array([0.22   , 0.16   , 0.315  , ..., 0.37125, 0.0325 , 0.04   ]), 'X', 'DRC Far North'))\n",
      "(20, (array([6.92930200e+03, 1.37947570e+04, 2.05072190e+04, ...,\n",
      "       6.13748295e+07, 6.14054660e+07, 6.14376831e+07]), array([0.05875, 0.04875, 0.08625, ..., 0.24875, 0.06125, 0.085  ]), '2R', 'Zambia'))\n",
      "(21, (array([  106585.182 ,   184446.54  ,   221036.9535, ..., 49335563.836 ,\n",
      "       49342660.191 , 49349478.1905]), array([0.0525 , 0.31875, 0.1425 , ..., 0.27375, 0.245  , 0.26625]), '2L', 'Zambia'))\n",
      "(22, (array([9.59644600e+03, 1.86722940e+04, 2.45557020e+04, ...,\n",
      "       5.29666158e+07, 5.30451037e+07, 5.30983105e+07]), array([0.05375, 0.32625, 0.11   , ..., 0.1725 , 0.0375 , 0.09875]), '3R', 'Zambia'))\n",
      "(23, (array([   68036.8035,    82886.502 ,   111535.469 , ..., 41931997.7705,\n",
      "       41943264.259 , 41951070.9215]), array([0.03375, 0.06375, 0.04875, ..., 0.26875, 0.27625, 0.4025 ]), '3L', 'Zambia'))\n",
      "(24, (array([4.78364350e+03, 2.33791065e+04, 3.78083210e+04, ...,\n",
      "       2.42617551e+07, 2.42873805e+07, 2.43315226e+07]), array([0.77   , 0.61125, 0.6875 , ..., 0.32625, 0.0325 , 0.0525 ]), 'X', 'Zambia'))\n",
      "(25, (array([6.92930200e+03, 1.37947570e+04, 2.05072190e+04, ...,\n",
      "       6.13748295e+07, 6.14054660e+07, 6.14376831e+07]), array([0.0725 , 0.03375, 0.095  , ..., 0.2925 , 0.2475 , 0.19875]), '2R', 'Burkina Faso'))\n",
      "(26, (array([  106585.182 ,   184446.54  ,   221036.9535, ..., 49335563.836 ,\n",
      "       49342660.191 , 49349478.1905]), array([0.04375, 0.645  , 0.31625, ..., 0.04125, 0.065  , 0.045  ]), '2L', 'Burkina Faso'))\n",
      "(27, (array([9.59644600e+03, 1.86722940e+04, 2.45557020e+04, ...,\n",
      "       5.29666158e+07, 5.30451037e+07, 5.30983105e+07]), array([0.0325 , 0.0725 , 0.0325 , ..., 0.09875, 0.02875, 0.0475 ]), '3R', 'Burkina Faso'))\n",
      "(28, (array([   68036.8035,    82886.502 ,   111535.469 , ..., 41931997.7705,\n",
      "       41943264.259 , 41951070.9215]), array([0.05375, 0.0725 , 0.04375, ..., 0.37125, 0.34125, 0.195  ]), '3L', 'Burkina Faso'))\n",
      "(29, (array([4.78364350e+03, 2.33791065e+04, 3.78083210e+04, ...,\n",
      "       2.42617551e+07, 2.42873805e+07, 2.43315226e+07]), array([0.265  , 0.11   , 0.19625, ..., 0.2025 , 0.03875, 0.0325 ]), 'X', 'Burkina Faso'))\n",
      "(30, (array([6.92930200e+03, 1.37947570e+04, 2.05072190e+04, ...,\n",
      "       6.13748295e+07, 6.14054660e+07, 6.14376831e+07]), array([0.0575 , 0.02875, 0.0525 , ..., 0.08375, 0.06   , 0.02875]), '2R', 'Central African Republic'))\n",
      "(31, (array([  106585.182 ,   184446.54  ,   221036.9535, ..., 49335563.836 ,\n",
      "       49342660.191 , 49349478.1905]), array([0.02875, 0.09625, 0.0375 , ..., 0.06   , 0.0575 , 0.0325 ]), '2L', 'Central African Republic'))\n",
      "(32, (array([9.59644600e+03, 1.86722940e+04, 2.45557020e+04, ...,\n",
      "       5.29666158e+07, 5.30451037e+07, 5.30983105e+07]), array([0.02625, 0.1275 , 0.0475 , ..., 0.07875, 0.0375 , 0.04625]), '3R', 'Central African Republic'))\n",
      "(33, (array([   68036.8035,    82886.502 ,   111535.469 , ..., 41931997.7705,\n",
      "       41943264.259 , 41951070.9215]), array([0.0325 , 0.0525 , 0.0375 , ..., 0.46375, 0.31375, 0.28875]), '3L', 'Central African Republic'))\n",
      "(34, (array([4.78364350e+03, 2.33791065e+04, 3.78083210e+04, ...,\n",
      "       2.42617551e+07, 2.42873805e+07, 2.43315226e+07]), array([0.2    , 0.18125, 0.3425 , ..., 0.5325 , 0.045  , 0.03875]), 'X', 'Central African Republic'))\n",
      "(35, (array([6.92930200e+03, 1.37947570e+04, 2.05072190e+04, ...,\n",
      "       6.13748295e+07, 6.14054660e+07, 6.14376831e+07]), array([0.07375, 0.03625, 0.04125, ..., 0.1175 , 0.1425 , 0.05375]), '2R', 'Cameroon'))\n",
      "(36, (array([  106585.182 ,   184446.54  ,   221036.9535, ..., 49335563.836 ,\n",
      "       49342660.191 , 49349478.1905]), array([0.02875, 0.11125, 0.07125, ..., 0.05375, 0.03375, 0.03875]), '2L', 'Cameroon'))\n",
      "(37, (array([9.59644600e+03, 1.86722940e+04, 2.45557020e+04, ...,\n",
      "       5.29666158e+07, 5.30451037e+07, 5.30983105e+07]), array([0.02625, 0.13875, 0.0475 , ..., 0.12875, 0.02875, 0.04   ]), '3R', 'Cameroon'))\n",
      "(38, (array([   68036.8035,    82886.502 ,   111535.469 , ..., 41931997.7705,\n",
      "       41943264.259 , 41951070.9215]), array([0.04   , 0.0525 , 0.05125, ..., 0.34125, 0.2625 , 0.21625]), '3L', 'Cameroon'))\n",
      "(39, (array([4.78364350e+03, 2.33791065e+04, 3.78083210e+04, ...,\n",
      "       2.42617551e+07, 2.42873805e+07, 2.43315226e+07]), array([0.1975 , 0.09625, 0.12625, ..., 0.29   , 0.03875, 0.0325 ]), 'X', 'Cameroon'))\n",
      "(40, (array([6.92930200e+03, 1.37947570e+04, 2.05072190e+04, ...,\n",
      "       6.13748295e+07, 6.14054660e+07, 6.14376831e+07]), array([0.0675 , 0.035  , 0.07625, ..., 0.20125, 0.13625, 0.075  ]), '2R', 'Tanzania'))\n",
      "(41, (array([  106585.182 ,   184446.54  ,   221036.9535, ..., 49335563.836 ,\n",
      "       49342660.191 , 49349478.1905]), array([0.02875, 0.32125, 0.1    , ..., 0.0575 , 0.0625 , 0.065  ]), '2L', 'Tanzania'))\n",
      "(42, (array([9.59644600e+03, 1.86722940e+04, 2.45557020e+04, ...,\n",
      "       5.29666158e+07, 5.30451037e+07, 5.30983105e+07]), array([0.03625, 0.26625, 0.04   , ..., 0.1275 , 0.04125, 0.05625]), '3R', 'Tanzania'))\n",
      "(43, (array([   68036.8035,    82886.502 ,   111535.469 , ..., 41931997.7705,\n",
      "       41943264.259 , 41951070.9215]), array([0.0325 , 0.04625, 0.02875, ..., 0.31625, 0.12375, 0.22   ]), '3L', 'Tanzania'))\n",
      "(44, (array([4.78364350e+03, 2.33791065e+04, 3.78083210e+04, ...,\n",
      "       2.42617551e+07, 2.42873805e+07, 2.43315226e+07]), array([0.17875, 0.0725 , 0.12375, ..., 0.1225 , 0.045  , 0.02625]), 'X', 'Tanzania'))\n",
      "(45, (array([6.92930200e+03, 1.37947570e+04, 2.05072190e+04, ...,\n",
      "       6.13748295e+07, 6.14054660e+07, 6.14376831e+07]), array([0.05625, 0.03625, 0.06   , ..., 0.1925 , 0.13625, 0.0975 ]), '2R', 'Uganda'))\n",
      "(46, (array([  106585.182 ,   184446.54  ,   221036.9535, ..., 49335563.836 ,\n",
      "       49342660.191 , 49349478.1905]), array([0.02625, 0.41875, 0.08125, ..., 0.06   , 0.04   , 0.0425 ]), '2L', 'Uganda'))\n",
      "(47, (array([9.59644600e+03, 1.86722940e+04, 2.45557020e+04, ...,\n",
      "       5.29666158e+07, 5.30451037e+07, 5.30983105e+07]), array([0.0325 , 0.22   , 0.0425 , ..., 0.0825 , 0.02875, 0.055  ]), '3R', 'Uganda'))\n",
      "(48, (array([   68036.8035,    82886.502 ,   111535.469 , ..., 41931997.7705,\n",
      "       41943264.259 , 41951070.9215]), array([0.04   , 0.05125, 0.02625, ..., 0.265  , 0.2625 , 0.14125]), '3L', 'Uganda'))\n",
      "(49, (array([4.78364350e+03, 2.33791065e+04, 3.78083210e+04, ...,\n",
      "       2.42617551e+07, 2.42873805e+07, 2.43315226e+07]), array([0.28875, 0.12875, 0.37125, ..., 0.31875, 0.055  , 0.0375 ]), 'X', 'Uganda'))\n"
     ]
    }
   ],
   "source": [
    "# Get the unique row and column variables\n",
    "row_var_levels = drc_metadata.group.unique()\n",
    "\n",
    "#set y labs to 2 dp\n",
    "def format_func(value, tick_number):\n",
    "    return f\"{value:.2f}\"\n",
    "\n",
    "# Initialize the figure and GridSpec\n",
    "fig = plt.figure(figsize=(sum(column_widths) * 25, len(row_var_levels) * 1.5))  # Plots half as high\n",
    "gs = GridSpec(len(row_var_levels), len(col_var_levels) + 1, width_ratios=list(column_widths) + [0.1])\n",
    "\n",
    "# Create the subplots\n",
    "for row_idx, row_val in enumerate(rowvars_reordered):\n",
    "    for col_idx, col_val in enumerate(col_var_levels):\n",
    "        row_colour = pop_code_cols.get(row_val, 'black')  # Default color if not specified\n",
    "\n",
    "        ax = fig.add_subplot(gs[row_idx, col_idx])\n",
    "        subset = selscan_df[(selscan_df['pop'] == row_val) & (selscan_df['chrom'] == col_val)]\n",
    "        sns.lineplot(data=subset, x='midpos', y='h1', ax=ax, linewidth=0.8, color=row_colour)\n",
    "\n",
    "        ax.fill_between(x=subset['midpos'], y1=subset['h1'], color=row_colour, alpha=0.1)  # Adjust alpha for transparency if needed\n",
    "\n",
    "\n",
    "        #Let's start by tinkering with the axes\n",
    "\n",
    "        #set max on y to be max h1 for each pop - this emphasises the sweep signal but we must note in the legend!\n",
    "        ax.set_ylim(0, 1)\n",
    "        ax.yaxis.set_major_formatter(FuncFormatter(format_func))\n",
    "        #set y max as ticks\n",
    "        ax.set_yticks([0, 1])\n",
    "        ax.set_ylabel(\"\")\n",
    "        plt.subplots_adjust(top=0.85) \n",
    "\n",
    "        #now we want to remove clutter as much as possible\n",
    "        #despine\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "        #remove x axes except bottom one\n",
    "        if row_idx < len(rowvars_reordered) - 1:\n",
    "            ax.xaxis.set_tick_params(which='both', bottom=False, top=False, labelbottom=False)  # Hide x-axis ticks and labels\n",
    "\n",
    "        # Remove y-axis for 2nd and 3rd columns\n",
    "        if col_idx > 0:\n",
    "            ax.set_ylabel('')\n",
    "            ax.yaxis.set_visible(False)\n",
    "\n",
    "        # Remove x-axis titles and replace with chromosome numbering\n",
    "        ax.set_xlabel('')\n",
    "        if row_idx == 10:\n",
    "            if col_idx == 0:\n",
    "                ax.set_xlabel(\"2R\", fontsize=24, fontweight='bold')\n",
    "            elif col_idx == 1:\n",
    "                ax.set_xlabel(\"2L\", fontsize=24, fontweight='bold')\n",
    "            elif col_idx == 2:\n",
    "                ax.set_xlabel(\"3R\", fontsize=24, fontweight='bold')  \n",
    "            elif col_idx == 3:\n",
    "                ax.set_xlabel(\"3L\", fontsize=24, fontweight='bold')\n",
    "            elif col_idx == 4:\n",
    "                ax.set_xlabel(\"X\", fontsize=24, fontweight='bold')\n",
    "\n",
    "        #add row (population) labels\n",
    "        row_ax = fig.add_subplot(gs[row_idx, -1])\n",
    "        row_ax.text(0.5, 0.5, row_val, va='center', ha='center', fontsize=18, rotation=0, transform=row_ax.transAxes)\n",
    "        row_ax.axis('off')\n",
    "\n",
    "        #now, let's annotate our plots\n",
    "\n",
    "        # #add text annotations and lines for IR genes\n",
    "        # if col_val == 'CM023248' and row_idx == 0:\n",
    "        #     # Cyp6\n",
    "        #     ax.fill_betweenx([cyp6_region[col_val]['y_min'], cyp6_region[col_val]['y_max']],\n",
    "        #         cyp6_region[col_val]['x_min'], cyp6_region[col_val]['x_max'],\n",
    "        #         color='gray', alpha=0.5)\n",
    "        #     ax.text(cyp6_region[col_val]['x_min'], 1, \"Cyp6\", fontsize=18, rotation=90)\n",
    "        #     # Ace1\n",
    "        #     ax.text(ace1_region[col_val]['x_min'], 1, \"Ace1\", fontsize=18, rotation=90)\n",
    "        #     ax.fill_betweenx([ace1_region[col_val]['y_min'], ace1_region[col_val]['y_max']],\n",
    "        #         ace1_region[col_val]['x_min'], ace1_region[col_val]['x_max'],\n",
    "        #         color='gray', alpha=0.5)\n",
    "\n",
    "        # elif col_val == 'CM023249' and row_idx == 0:\n",
    "        #     #Vgsc\n",
    "        #     ax.fill_betweenx([vgsc_region[col_val]['y_min'], vgsc_region[col_val]['y_max']],\n",
    "        #         vgsc_region[col_val]['x_min'], vgsc_region[col_val]['x_max'],\n",
    "        #         color='gray', alpha=0.5)\n",
    "        #     ax.text(vgsc_region[col_val]['x_min'], 1, \"Vgsc\", fontsize=18, rotation=90)\n",
    "\n",
    "        #     #Gste\n",
    "        #     ax.fill_betweenx([gste_region[col_val]['y_min'], gste_region[col_val]['y_max']],\n",
    "        #         gste_region[col_val]['x_min'], gste_region[col_val]['x_max'],\n",
    "        #         color='gray', alpha=0.5)\n",
    "        #     ax.text(gste_region[col_val]['x_min'], 1, \"Gste\", fontsize=18, rotation=90)\n",
    "\n",
    "        #     #Rdl\n",
    "        #     ax.fill_betweenx([rdl_region[col_val]['y_min'], rdl_region[col_val]['y_max']],\n",
    "        #         rdl_region[col_val]['x_min'], rdl_region[col_val]['x_max'],\n",
    "        #         color='gray', alpha=0.5)\n",
    "        #     ax.text(rdl_region[col_val]['x_min'],1, \"Rdl\", fontsize=18, rotation=90)\n",
    "\n",
    "        #     #Coeae\n",
    "        #     ax.fill_betweenx([carboxyl_cluster[col_val]['y_min'], carboxyl_cluster[col_val]['y_max']],\n",
    "        #                     carboxyl_cluster[col_val]['x_min'], carboxyl_cluster[col_val]['x_max'],\n",
    "        #                     color='gray', alpha=0.5)\n",
    "        #     ax.text(carboxyl_cluster[col_val]['x_min'], 1, \"Coeae\", fontsize=18, rotation=90)\n",
    "\n",
    "        #     #CoeJH\n",
    "        #     ax.fill_betweenx([COEJHE5E_region[col_val]['y_min'], COEJHE5E_region[col_val]['y_max']],\n",
    "        #                     COEJHE5E_region[col_val]['x_min'], COEJHE5E_region[col_val]['x_max'],\n",
    "        #                     color='gray', alpha=0.5)\n",
    "        #     ax.text(COEJHE5E_region[col_val]['x_min'], 1, \"Coejh\", fontsize=18, rotation=90)\n",
    "\n",
    "        # elif col_val == 'CM023250' and row_idx == 0:\n",
    "        #     #Cyp9k1\n",
    "        #     ax.fill_betweenx([cyp9_region[col_val]['y_min'], cyp9_region[col_val]['y_max']],\n",
    "        #         cyp9_region[col_val]['x_min'], cyp9_region[col_val]['x_max'],\n",
    "        #         color='gray', alpha=0.5)\n",
    "        #     ax.text(cyp9_region[col_val]['x_min'], 1, \"Cyp9k1\", fontsize=18, rotation=90)\n",
    "\n",
    "        #     #Diagk\n",
    "        #     ax.fill_betweenx([diagk[col_val]['y_min'], diagk[col_val]['y_max']],\n",
    "        #         diagk[col_val]['x_min'], diagk[col_val]['x_max'],\n",
    "        #         color='gray', alpha=0.5)\n",
    "        #     ax.text(diagk[col_val]['x_min'], 1, \"Diagk\", fontsize=18, rotation=90)\n",
    "\n",
    "#Add master Y labels\n",
    "fig.text(0.08, 0.5, 'H1', va='center', rotation='vertical', fontweight='bold', fontsize=18)\n",
    "\n",
    "plt.subplots_adjust(wspace=0, hspace=0.2)  # Reduce column and row padding\n",
    "\n",
    "fig.savefig('../figures/h1.svg')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "#save as svg"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "malariagen",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
